libvirt/tests/src/cpu/diagnose_9c_forward.py

```
@@ -0,0 +1,129 @@
+import logging
+import time
+
+from avocado.core.exceptions import TestFail
+from virttest.libvirt_xml.vm_xml import VMXML
+from virttest.utils_misc import cmd_status_output, wait_for
+
+
+DEACTIVATE_FORWARDING = 0
+FORWARD_COUNTER_TIMEOUT = 360
+CHECK_INTERVAL = 60
+
+
+def update_kvm_parameter(hz):
+    """
+    Updates kvm paramter for forwarding
+
+    :param hz: diag9c_forwarding_hz
+    """
+
+    cmd = "echo %s > /sys/module/kvm/parameters/diag9c_forwarding_hz" % hz
+    err, out = cmd_status_output(cmd, shell=True)
+    if err:
+        raise OSError("Could not update parameter. %s" % out)
+
+
+def start_locktorture(session):
+    """
+    Starts spin lock torture test to cause spinlocks in guest
+
+    :param session: guest session
+    """
+    logging.debug("session is %s" % session)
+    cmd = "modprobe locktorture torture_type=spin_lock"
+    err, out = cmd_status_output(cmd, shell=True, session=session)
+    if err:
+        raise OSError("Could not start locktorture. %s" % out)
+
+
+def confirm_forwarding(hz):
+    """
+    Confirm if forwarding is happening or not depending on set
+    hz parameter
+
+    :param hz: if 0 expect no forwarding, if positive expect forward
+    """
+
+    if int(hz) > 0:
+        wait_for_forward_count_to_be_non_zero(FORWARD_COUNTER_TIMEOUT)
+    else:
+        confirm_forward_count_is_still_zero_after(CHECK_INTERVAL)
+
+
+def confirm_forward_count_is_still_zero_after(interval):
+    """
+    Wait for a certain time interval and then confirm
+    that the counter didn't go up
+    """
+    time.sleep(interval)
+    err, out = get_forward_value()
+    if err or int(out) != 0:
+        raise TestFail("Counter wasn't zero although forward"
```

As a common rule, we do use test.fail() instead of raising TestFail exception directly in this repo.

---

@dzhengfy I prefer to raise the exception here directly because then I don't have to pass test to the top level function and the top level function can be more easily moved, e.g. to avocado-vt if found useful for other tests. What do you think?